File: Medium/394. Decode String/Solution.java
class Solution {
    public String decodeString(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == ']') {
                StringBuilder text = new StringBuilder();
                char cur = stack.pop();
                while (cur != '[') {
                    text.insert(0, cur);
                    cur = stack.pop();
                }

                StringBuilder num = new StringBuilder();
                cur = stack.pop();

                while (Character.isDigit(cur)) {
                    num.insert(0, cur);
                    if (stack.isEmpty())
                        break;
                    cur = stack.pop();
                }

                if (!Character.isDigit(cur))
                    stack.push(cur);

                int n = Integer.parseInt(num.toString());
                String str = text.toString().repeat(n);

                for (char c1 : str.toCharArray())
                    stack.push(c1);
            } else {
                stack.push(c);
            }
        }

        StringBuilder res = new StringBuilder();
        for (char c : stack)
            res.append(c);

        return res.toString();
    }
}

File: Medium/394. Decode String/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function (s) {
    let stack = [];

    for (const c of s) {
        if (c == "]") {
            let txt = "";
            let cur = stack.pop();
            while (cur != "[") {
                txt = cur + txt;
                cur = stack.pop();
            }
            let num = "";
            cur = stack.pop();
            while (!Number.isNaN(+cur)) {
                num = cur + num;
                cur = stack.pop();
            }
            stack.push(cur);
            stack.push(txt.repeat(+num));
        } else {
            stack.push(c);
        }
    }

    return stack.join("");
};